tools/depcheck/pkg/cmd/trace.go

```diff
@@ -0,0 +1,293 @@
+	return cmd
+}
+
+func (o *TraceImportsOpts) Complete() error {
+	o.Roots = ensureRecursiveRootFormat(o.Roots)
+	return nil
+}
+
+func ensureRecursiveRootFormat(roots []string) []string {
+	parsedRoots := []string{}
+	for _, root := range roots {
+		if strings.HasSuffix(root, "...") {
+			parsedRoots = append(parsedRoots, root)
+			continue
+		}
+
+		slash := ""
+		if string(root[len(root)-1]) != "/" {
+			slash = "/"
+		}
+		root = strings.Join([]string{root, slash, "..."}, "")
+		parsedRoots = append(parsedRoots, root)
+	}
+
+	return parsedRoots
+}
+
+func (o *TraceImportsOpts) Validate() error {
+	if len(o.Roots) == 0 {
+		return errors.New("at least one root package must be provided")
+	}
+	if len(o.OutputFormat) != 0 && o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	return nil
+}
+
+type Node struct {
+	id         int
+	uniqueName string
+}
+
+func (n Node) ID() int {
+	return n.id
+}
+
+type PackageImports struct {
+	Imports []string
+}
+
+type Package struct {
+	ImportPath string
+	Imports    []string
+}
+
+type PackageList struct {
+	Packages []Package
+}
+
+func (p *PackageList) filterDeps(pkg *Package) {
+	validImports := []string{}
+	for _, dep := range pkg.Imports {
+		for _, valid := range validDepPrefixes {
+			if strings.HasPrefix(dep, valid) {
+				validImports = append(validImports, dep)
+				break
+			}
+		}
+	}
+
+	pkg.Imports = validImports
+}
+
+// TODO: provide a way for consumers of this command
+// to specify a differentiation between vendored and
+// non-vendored dependencies
+func (p *PackageList) FilteredAdd(pkg Package) {
+	if !strings.Contains(pkg.ImportPath, "/vendor/") {
+		p.filterDeps(&pkg)
+		p.Add(pkg)
+		return
+	}
+
+	for _, valid := range validPackages {
+		if strings.Contains(pkg.ImportPath, valid) {
+			p.filterDeps(&pkg)
+			p.Add(pkg)
+			return
+		}
+	}
+}
+
+func (p *PackageList) Add(pkg Package) {
+	p.Packages = append(p.Packages, pkg)
+}
+
+// Run execs `go list` on all package entrypoints specified through --root.
+// Each package's ImportPath and non-transitive (immediate) imports are
+// filtered and aggregated. A package is filtered based on whether its ImportPath
+// matches an accepted pattern defined in the set of validPackages.
+// Each aggregated package becomes a node in a generated dependency graph.
+// An edge is created between a package and each of its Imports.
+func (o *TraceImportsOpts) Run() error {
+	args := []string{"list", "--json"} // "./pkg/...", "./vendor/...", "./cmd/.."
+	golist := exec.Command("go", append(args, o.Roots...)...)
+
+	r, w := io.Pipe()
+	golist.Stdout = w
+	golist.Stderr = os.Stderr
+
+	pkgs := &PackageList{}
+	go func(list *PackageList) {
+		decoder := json.NewDecoder(r)
+		for {
+			var pkg Package
+			err := decoder.Decode(&pkg)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			}
+
+			list.FilteredAdd(pkg)
```

Comment:
I don't think you should filter your input here.  Filter when creating your graph instead